File: src/main/java/javaclasses/HeapPriorityQueue/MergeKSortedLists.java
/*
 * *
 *  * Merge K Sorted Lists.java
 *  * Created by Rafsan Ahmad on 12/7/21, 9:05 PM
 *  * Copyright (c) 2022 . All rights reserved.
 *
 */

package javaclasses.HeapPriorityQueue;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

import javaclasses.LinkedList.ListNode;

public class MergeKSortedLists {
    //Leetcode 23
    /*
    * You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.
    Merge all the linked-lists into one sorted linked-list and return it.

     Example 1:

    Input: lists = [[1,4,5],[1,3,4],[2,6]]
    Output: [1,1,2,3,4,4,5,6]
    Explanation: The linked-lists are:
    [
      1->4->5,
      1->3->4,
      2->6
    ]
    merging them into one sorted list:
    1->1->2->3->4->4->5->6
    Example 2:

    Input: lists = []
    Output: []
    Example 3:

    Input: lists = [[]]
    Output: []*/

    //Runtime: O(nlogn)
    public ListNode mergeKListsNaive(ListNode[] lists) {
        ListNode resultNode = new ListNode();
        ListNode tmp = resultNode;
        List<Integer> values = new ArrayList<>();

        for (int i = 0; i < lists.length; i++) {
            ListNode current = lists[i];
            while (current != null) {
                values.add(current.val);
                current = current.next;
            }
        }
        Collections.sort(values);
        for (int i = 0; i < values.size(); i++) {
            tmp.next = new ListNode(values.get(i));
            tmp = tmp.next;
        }
        return resultNode.next;
    }

    //Using Min-heap
    //Time complexity: O(nlog(k))
    //Space complexity: O(k)
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null) return null;

        PriorityQueue<ListNode> queue = new PriorityQueue(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                return o1.val - o2.val;
            }
        });

        // push the first node of each list into the min-heap
        for (ListNode node : lists) {
            if (node != null) queue.add(node);
        }

        // take two pointers, head and tail, where the head points to the first node
        // of the output list and tail points to its last node
        ListNode head = null, last = null;

        // run till min-heap is empty
        while (!queue.isEmpty()) {
            // extract the minimum node from the min-heap
            ListNode min = queue.poll();

            // add the minimum node to the output list
            if (head == null) {
                head = last = min;
            } else {
                last.next = min;
                last = min;
            }

            // take the next node from the "same" list and insert it into the min-heap
            if (min.next != null) queue.add(min.next);
        }

        // return head node of the merged list
        return head;
    }

    public void printList(ListNode node) {
        while (node != null) {
            System.out.print(node.val + " —> ");
            node = node.next;
        }
        System.out.print("null");
    }

    public static void main(String[] args) {
        MergeKSortedLists list = new MergeKSortedLists();
        int k = 3;    // total number of linked lists

        // an array to store the head nodes of the linked lists
        ListNode[] lists = new ListNode[k];

        lists[0] = new ListNode(1);
        lists[0].next = new ListNode(5);
        lists[0].next.next = new ListNode(7);

        lists[1] = new ListNode(2);
        lists[1].next = new ListNode(3);
        lists[1].next.next = new ListNode(6);
        lists[1].next.next.next = new ListNode(9);

        lists[2] = new ListNode(4);
        lists[2].next = new ListNode(8);
        lists[2].next.next = new ListNode(10);

        // Merge all lists into one
        ListNode head = list.mergeKLists(lists);
        list.printList(head);
    }
}

File: src/main/java/kotlinclasses/HeapPriorityQueue/MergeKSortedLists.kt
/*
 * *
 *  * Merge k Sorted Lists.kt
 *  * Created by Rafsan Ahmad on 8/28/25, 10:02PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.HeapPriorityQueue

import kotlinclasses.LinkedList.ListNode
import java.util.PriorityQueue

class MergeKSortedLists {
    //https://leetcode.com/problems/merge-k-sorted-lists/description/
    /*You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

Merge all the linked-lists into one sorted linked-list and return it.

Example 1:
Input: lists = [[1,4,5],[1,3,4],[2,6]]
Output: [1,1,2,3,4,4,5,6]
Explanation: The linked-lists are:
[
  1->4->5,
  1->3->4,
  2->6
]
merging them into one sorted linked list:
1->1->2->3->4->4->5->6

Example 2:
Input: lists = []
Output: []

Example 3:
Input: lists = [[]]
Output: []


Constraints:
k == lists.length
0 <= k <= 10^4
0 <= lists[i].length <= 500
-10^4 <= lists[i][j] <= 10^4
lists[i] is sorted in ascending order.
The sum of lists[i].length will not exceed 10^4.*/

    /*Time Complexity:
Let N = total number of nodes across all lists.
Let k = number of lists.
Each poll() and add() on heap costs O(log k).
We perform N such operations.
Total: O(N log k)

Space Complexity:
Heap stores at most k nodes → O(k)
Result list reuses existing nodes → O(1) extra
Overall: O(k)*/
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        if (lists.isEmpty()) return null

        // Min-heap (priority queue) based on node values
        val queue: PriorityQueue<ListNode> = PriorityQueue { a, b -> a.`val` - b.`val` }

        // Dummy node to simplify result construction
        val result = ListNode(0)
        var tail: ListNode? = result

        // Step 1: Push the head of each list into the min-heap
        for (i in lists.indices) {
            if (lists[i] != null)
                queue.add(lists[i])
        }

        // Step 2: Process heap until empty
        while (!queue.isEmpty()) {
            // Extract the smallest node among the k lists
            val currNode = queue.poll()

            // Append it to the result list
            tail?.next = currNode
            tail = tail?.next

            // KEY STEP:
            // If current node has a next, push it into heap.
            // This ensures that at any time, heap contains at most one "active" node from each list.
            // Hence, the heap always knows the global minimum among remaining nodes.
            if (currNode.next != null)
                queue.add(currNode.next)
        }

        // Skip dummy node and return merged head
        return result.next
    }
}

fun main() {
    val obj = MergeKSortedLists()
    // Build test input: lists = [[1,4,5],[1,3,4],[2,6]]
    val l1 = ListNode(1).apply {
        next = ListNode(4).apply {
            next = ListNode(5)
        }
    }
    val l2 = ListNode(1).apply {
        next = ListNode(3).apply {
            next = ListNode(4)
        }
    }
    val l3 = ListNode(2).apply {
        next = ListNode(6)
    }

    val lists = arrayOf<ListNode?>(l1, l2, l3)
    val merged = obj.mergeKLists(lists)

    // Print output
    var curr = merged
    print("Output: [")
    while (curr != null) {
        print(curr.`val`)
        curr = curr.next
        if (curr != null) print(",")
    }
    println("]")
}